2-level multi-signing implementation [RIPD-182]

scottschurr opened the pull request:
This set of commits brings support for 2-level multi-signing into
the code base.  The capability is (typically) compiled out, since
multi-signing is not feature complete.

Here's a guide to the commits:

 o Improvements to protocol serialization: This commit pulled out a
   few general improvements to SField and STObject.  It was
   previously a separate pull request.  Vinnie and Nik suggested
   changes that I have introduced in this version of the commit.

 o Multisign whitespace: changes to white space, adding or removing
   virtuals.  Mostly trying to move noise out of the other commits.

 o SignerSetList transaction: this commit fully introduces the
   SignerListSet transaction.  New, since the last m-of-n review,
   is the addition of InnerObjectFormats.  Code related to the use
   of InnerObjectFormats will need the closest review.

 o get_signingaccount RPC command: this commit fully introduces
   the get_signing_account RPC command, which returns individual
   multi-signatures.  New, since the last m-of-n review, is support
   for 2-level signing.  That change impacted much of the previously
   reviewed code.

 o 2-level transaction multi-signatures: this commit finally allows
   for 2-level multi-signatures.  Again, much of the previously
   reviewed code was touched in order to adjust from single- to
   two-level multi-signing.

The user availability of the code is managed by a #define in
BeastConfig.  Once amendments are available I'll switch the code
over to be managed by an amendment.

@nbougalis, @JoelKatz

sublimator commented on src/ripple/app/ledger/LedgerEntrySet.cpp:
```diff
@@ -1024,52 +1024,66 @@ uint256 LedgerEntrySet::getNextLedgerIndex (
-            " owner count is already 0!";
-        return;
+            " owner count set below 0!";
+        new_count = 0;
```
Something to ponder: You know, if we had an assert(false) here for debug mode, we would have caught a recent bug with the TicketCancel transactor, where the wrong account's owner count was decremented.   It's a cheap thing to do, that doesn't require all that much consideration, compared to production error handling / exceptions etc.  The log level says lsFATAL.

scottschurr replied:
I'm more than happy to add an `assert` here.  Would that help much?  I'm under the impression that a debug rippled has trouble keeping up with the network traffic.  On the other hand it can't hurt, since the production servers are all running optimized code.  Thoughts?

sublimator replied:
I can say with 95% confidence the existing npm tests would have triggered the assert and the bug would have been found sooner.  The tests DID exercise this code, though failed to exhaustively assert expected behaviour for all the changes.

scottschurr replied:
95% is good enough for me.  I added the assert.

sublimator commented on src/ripple/app/transactors/README.md:
```diff
@@ -0,0 +1,217 @@
+For accounts which are designated as multi-sign, there must be a list which
+specifies which accounts are authorized to sign and the quorum threshold.
+
+ - Each authorized account has a weight. The sum of the weights of the signers is used to determine whether a given set of signatures is sufficient for a quorum.
```
Rest of the doc is using hard wraps

scottschurr replied:
This certainly results from my lack of knowledge of markdown.  I couldn't find a way to get the bullets to work if I wrapped these lines.

sublimator commented on src/ripple/protocol/impl/TxFormats.cpp:
```diff
@@ -84,6 +84,13 @@ TxFormats::TxFormats ()
+
+    // The SignerEntries are optional because a SignerList is deleted by
+    // setting the SignerQuorum to zero and omitting SignerEntries.
+    add ("SignerListSet", ttSIGNER_LIST_SET)
```
I guess this bird has already flown, (tickets/offers etc) but using SignerListSet and SetSignerList in the same codebase seems unfortunate. Makes it a touch harder to grep the code/logs.

scottschurr replied:
I agree.  Arthur has argued that the `Set` transaction types should all have the `Set` at the end for consistency.  But for files and constants I picked names that were consistent with those in the codebase neighborhood, which meant that the `Set` went at the front for those.  It felt awkward at the time (as it still does).

However when I look at
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/protocol/impl/TxFormats.cpp#L55
it appears that names of transactions may already be inconsistent on this point.  It's true that I'm seeing `AccountSet` and `TrustSet`, but I'm also seeing `SetRegularKey` and `SetFee`.  I think `SetSignerList` would be easier for an English reader to comprehend and would be consistent with the file and object names.  But I'm not ready to go to the mat over this one.

sublimator replied:
Consistently inconsistent!

sublimator commented on src/ripple/app/transactors/SetSignerList.cpp:
```diff
@@ -0,0 +1,322 @@
+    std::uint64_t const newReserve =
+        mEngine->getLedger ()->getReserve (oldOwnerCount + addedOwnerCount);
+
+    if (mPriorBalance.getNValue () < newReserve)
```
There's not much in it, but do we want `mSourceBalance` here, which seem to be the Balance - Fee ?

sublimator replied:
Ah, this is so we can use reserve to pay Fee eh?

scottschurr replied:
Great question, and I'd forgotten this detail.  I modeled this implementation off of `TicketCreate`, which allows the fee to come out of the reserve.  But I should certainly add a comment to that effect.

sublimator commented on src/ripple/app/transactors/SetSignerList.cpp:
```diff
@@ -0,0 +1,322 @@
+        }
+
+        // Don't verify that the signer accounts exist.  Verifying them is
+        // expensive and they may not exist yet due to network phenomena.
```
+ phantom accounts

sublimator commented on src/ripple/app/transactors/SetSignerList.cpp:
```diff
@@ -0,0 +1,322 @@
+    // We can get away with that because lists are not adjusted incrementally;
+    // we add or remove an entire list.
+
+    // The wiki currently says (December 2014) the reserve should be
```
For others:  https://wiki.ripple.com/Multisign#Fees_2

sublimator commented on src/ripple/app/transactors/Transactor.cpp:
```diff
@@ -302,4 +284,356 @@ TER Transactor::apply ()
+    {
+        journal.trace <<
+            "applyTransaction: Invalid: Not a multi-signing account.";
+        ret.ter = temBAD_AUTH_MASTER;
```
temBAD_AUTH_MASTER isn't a very useful error message in this case:  `Auth for unclaimed account needs correct master key.`

sublimator replied:
With tem*, or more to the point, anything that isn't tec/tes, and doesn't need to be serialized, there's not as much need to be conservative about introducing new error codes.

scottschurr replied:
Yeah, I need to re-think this a bit.  What we have here is code that preserves the original behavior if an empty `SigningPublicKey` were passed.  But the return code makes no sense in the context of a second-level signer.

I'm guessing the best solution is to give up on preserving the old behavior if someone passed an empty `SigningPublicKey`.  Probably a new `tefNOT_MULTI_SIGNING` code would be the right way to go.  That makes sense in all contexts except the one where someone passes an empty `SigningPublicKey` and multi-signing is deployed on the network but not enabled by an amendment.  That's not a case worth worrying about.

Does anyone have heartburn with me adding `tefNOT_MULTI_SIGNING`?  It will probably need tuning once we add access control flags.

scottschurr replied:
I've added a `tefNOT_MULTI_SIGNING` code.  I picked a `tef` code based on this comment in TER.h:
```
    // Causes:
    // - Transaction cannot succeed because of ledger state.
```
I'm using that new code here because the ledger doesn't contain a signer list for the account.

sublimator commented on src/ripple/protocol/impl/STTx.cpp:
```diff
@@ -272,6 +274,196 @@ STTx::getMetaSQL (Serializer rawTxn,
+
+    // SigningFors must be in sorted order by AccountID.
+    RippleAddress lastSigningForID;
+    lastSigningForID.setAccountID ("");
```
Empty string is a valid way to get a zero account?

scottschurr replied:
Yup.  Thank heaven that I have access to the source for `RippleAddress` or I'd never have figured it out.  https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/protocol/impl/RippleAddress.cpp#L418.  I hate that class.

sublimator replied:
There's always this:  `void RippleAddress::setAccountID (Account const& hash160)` 
Would `lastSigningForID.setAccountID({0})` work?

scottschurr replied:
@sublimator: I think I tried initializing with `0` and failed.  `RippleAddress::setAccountID (std::string)` has special handling for an empty string.  See here:
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/protocol/impl/RippleAddress.cpp#L422
As I recall, if I `setAccountID(0)`, the resulting RippleAddress is not valid.

sublimator commented on src/ripple/app/transactors/SetSignerList.cpp:
```diff
@@ -0,0 +1,322 @@
+        }
+    }
+
+    // Make sure there are no duplicate signers.
```
Do we not want to check that each entry in the SignerEntries is ordered by sfAccount ?  i.e. check that it's canonical?

scottschurr replied:
This is another great question that I didn't know enough to ask several months ago.  As it stands, the `SignerEntries` always go into the ledger in a canonical form -- I sort the list once I get it.  But I don't insist that it arrive in canonical form.  I have no idea about the impact of not enforcing that the wire form is sorted.

That is different from the behavior of multi-signatures on transactions.  I *do* insist that they are sorted.  Although, as a convenience, the `submit_multisigned` RPC command does that sorting for a caller.

If David likes the idea of requiring that the `SignerEntries` arrive sorted I'm happy to implement that change.

scottschurr replied:
I checked in with David.  He's not particularly worried whether the `SignerEntries` in the transaction are sorted.  Not requiring them to be sorted makes it easier for less sophisticated clients.  So I'll leave this as it is.

sublimator replied:
Makes sense.

vinniefalco replied:
Whether they are sorted or not, they will be sorted during serialization for the purpose of signature verification. So I guess it doesn't matter (correct me if I'm wrong here).

scottschurr replied:
They do get sorted before going into the ledger.  So the ledger format has a canonical form.  David's not worried about them arriving in sorted order, I checked with him.

scottschurr commented:
I've rebased against the latest develop and folded in my fixes from sublimator (thanks Niq).

nbougalis commented:
Scott has already fixed all my concerns. This gets a  :+1: from me.

scottschurr commented:
I folded the `sign_for` name change into the appropriate commit.

vinniefalco commented:
This changes tx processing?

scottschurr commented:
@vinniefalco: It changes transaction processing in the sense that the order of some checks has changed.  So a transaction with an error might return a different error code than it previously did.  If `RIPPLE_ENABLE_MULTI_SIGN` is `0` then there should be no functional transaction changes (modulo bugs).

nbougalis commented:
I know that this has been sitting on the queue for a while now, but should we wait post 0.28 to merge it?

scottschurr commented:
@nbougalis Yes.  I'm under the impression that 0.28 is close to ready.  And we don't yet know which bugs David will find.  I think it probably makes sense to do the merge after 0.28 so it spends maximal time burning in on the development branch before going main stream.

scottschurr commented:
Hey folks, I rebased, did a very minor fold, and added the m-of-n unit tests to this pull request.  Nik and Niq, it would be great if you'd look over the unit test commit.  David, I'm afraid you have the full slog in front of you.  Thanks and good luck.

sublimator commented on src/ripple/app/transactors/test/MultiSign.test.cpp:
```diff
@@ -0,0 +1,1611 @@
+            STTx tx = getSignerListSetTx (alice, aliceSigners, 0);
+            singleSign (tx, alice);
+            ledger.applyGoodTransaction (tx);
+             aliceBalanceCheck (stdFee);
```
It would be nice to check things like the OwnerCount has decremented symmetrically

scottschurr replied:
Great idea.  I'm adding the OwnerCount check right now.  Are there other "things like" OwnerCount I should be checking as well?

JoelKatz commented:
Okay, I've been over all the code. Other than the comments I made, it all looks good. We need to get it rebased though. If it will make the rebasing easier, IMO, it can be squashed.

scottschurr commented:
I believe I've fixed everything except for one question: what to do when the same entity signs both at first and second level.

https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/protocol/impl/STTx.cpp#L339

The current implementation rejects the transaction, since it is signed twice by the same signer.  Another possibility, raised by David, would be to allow the transaction but only add the entity's weight toward the quorum once.

Once we pick the preferred solution to that issue I'll push that as a fold as well.

JoelKatz commented:
@scottschurr  Oh, I see. Perhaps I misunderstood. If the same account signs two different ways, and the only possible choices are to make this illegal or to count both the same as just one, making this illegal is probably better. And it's obvious that it's illegal without consulting the ledger, so it can be considered malformed. Then leave it as is.

scottschurr commented:
@JoelKatz Thanks.  I'll leave the code as is, but I'll expand the comment so it is easier for folks to see what we're preventing and why.

sublimator commented:
I know ripple-lib for example uses a valid ledger range of only 3 ledgers by default, and we recommend to always use LastLedgerSequence when submitting transactions. How's this going to play out with multi signing? LastLedgerSequence is quite important for determining a transactions final state.

scottschurr commented:
@sublimator My understanding regarding LastLedgerSequence is that it will probably not be useful in most multi-signing situations.  The unsigned transaction must be fully formed (including fields like Fee and LastLedgerSequence) before the out-of-ledger signing process can begin.  Since the signing process takes an indeterminate amount of time, it would be difficult to predict a useful LastLedgerSequence.

The best alternative I've seen so far is to use a Ticket for the transaction, since the Ticket can be cancelled if the transaction gets caught in limbo.  Other ideas are welcome and encouraged.

sublimator commented:
>>> it would be difficult to predict a useful LastLedgerSequence.

Exactly.

We still don't have ticket usage support yet huh, just disabled
create/cancel? Having to cancel the ticket (or use a noop
AccountSet, `Sequence plug` transaction) if you aren't sure of the result,
doesn't sound like much fun.

scottschurr commented on src/ripple/rpc/impl/Handler.cpp:
```diff
@@ -130,7 +130,9 @@ HandlerTable HANDLERS({
     {   "random",               byRef (&doRandom),              Role::USER,  NO_CONDITION     },
     {   "ripple_path_find",     byRef (&doRipplePathFind),      Role::USER,  NEEDS_CURRENT_LEDGER  },
     {   "sign",                 byRef (&doSign),                Role::USER,  NO_CONDITION     },
+    {   "sign_for",             byRef (&doSignFor),             Role::USER,    NO_CONDITION     },
```
Thanks @sublimator.  The most recent [FOLD] commit incorporates this change.

scottschurr commented:
Rebased against the latest develop.  Commit d8b10ec (marked `[FOLD] Move some error checks...`) changes the `SetSignerList::preCheck()` to be in line with recent changes to the other Transactors.  It certainly deserves some review attention at some point.

scottschurr commented:
m-of-n is now rebased against the latest develop branch.  I'm leaving the `Hold` label in place until @nbougalis and @sublimator have had a chance to look over the rebased code.  I'm not expecting @JoelKatz to look over the code again, although he's welcome to if he wants to.  If Nik and Niq say it's okay I'll put the `Passed` label on it.

vinniefalco commented:
Unit test sources need to go in `.../tests/...`

scottschurr commented:
Unit tests are moved and Nik's last two comments addressed.
